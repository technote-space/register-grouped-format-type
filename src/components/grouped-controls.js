const { createSlotFill } = wp.components;

/**
 * @param {string} group group
 * @returns {*} slot fill
 * @constructor
 */
const GroupedControls = group => createSlotFill('GroupedControls.' + group);

export default GroupedControls;
