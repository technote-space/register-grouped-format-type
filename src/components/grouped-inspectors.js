const {createSlotFill} = wp.components;

/**
 * @param {string} group group
 * @returns {*} slot fill
 * @constructor
 */
const GroupedInspectors = group => createSlotFill('GroupedInspectors.' + group);

export default GroupedInspectors;
