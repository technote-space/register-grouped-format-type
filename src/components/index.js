import GroupedControls from './grouped-controls';
import GroupedInspectors from './grouped-inspectors';
import ToolbarDropdown from './toolbar-dropdown';
import InspectorComponent from './inspector-component';

export {GroupedControls, GroupedInspectors, ToolbarDropdown, InspectorComponent};
