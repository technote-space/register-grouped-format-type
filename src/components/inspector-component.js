import { Helpers } from '../utils';

const { getEditor } = Helpers;
const { InspectorControls } = getEditor();
const { PanelBody } = wp.components;
const { getContrastChecker } = Helpers;

/**
 * @param {Array} fills fills
 * @param {object} setting setting
 * @param {object} args args
 * @returns {null|*} inspector
 */
const createInspectorComponent = ( fills, setting, args ) => {
	const activeFills = fills.filter( ( [ { props } ] ) => ! props.isDisabled );
	const additionalInspectors = setting.additionalInspectors;
	if ( ! activeFills.length && ! additionalInspectors.length ) {
		return null;
	}

	return <InspectorControls>
		<PanelBody
			{ ...setting.inspectorSettings }
		>
			{ activeFills }
			{ setting.useContrastChecker && getContrastChecker( activeFills, args ) }
			{ additionalInspectors.map( component => component( args, setting ) ) }
		</PanelBody>
	</InspectorControls>;
};

/**
 * @param {object} Slot Slot
 * @param {object} setting setting
 * @param {object} args args
 * @returns {*} inspector
 * @constructor
 */
const InspectorComponent = ( Slot, setting, args ) => <Slot>
	{ fills => ! fills.length ? null : createInspectorComponent( fills, setting, args ) }
</Slot>;

export default InspectorComponent;
