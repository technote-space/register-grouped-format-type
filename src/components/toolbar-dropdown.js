import classnames from 'classnames';
import { Components } from '@technote-space/gutenberg-utils';

const { BlockFormatControls } = wp.blockEditor;
const { Toolbar, IconButton, NavigableMenu } = wp.components;

/**
 * @param {Array} fills fills
 * @param {object} setting setting
 * @returns {*} dropdown
 */
const createDropdown = ( fills, setting ) => {
	const controls = fills.map( ( [ { props } ] ) => props );
	const isActive = !! controls.filter( control => control.isActive ).length;
	const isDisabled = ! controls.filter( control => ! control.isDisabled ).length;
	const onClick = ( props, control ) => event => {
		event.stopPropagation();
		props.onClose();
		if ( control.onClick ) {
			control.onClick();
		}
	};
	const { label, menuLabel, menuClassName } = setting;
	return <Components.DropdownButton
		{ ...setting }
		isActive={ isActive }
		isDisabled={ isDisabled }
		renderContent={ ( props ) => <NavigableMenu
			className={ classnames( 'components-dropdown-menu__menu', menuClassName ) }
			role="menu"
			aria-label={ menuLabel || label }
		>
			{ controls.map( ( control, index ) => <IconButton
				key={ index }
				onClick={ onClick( props, control ) }
				className={ classnames( 'components-dropdown-menu__menu-item', {
					'is-active': control.isActive,
				} ) }
				icon={ control.icon }
				role="menuitem"
				disabled={ control.isDisabled }
			>
				{ control.title }
			</IconButton> ) }
		</NavigableMenu> }
	/>;
};

/**
 * @param {Array} fills fills
 * @param {object} setting setting
 * @returns {*} component
 */
const createComponent = ( fills, setting ) => fills.length > 1 ? createDropdown( fills, setting ) : fills[ 0 ]; /* eslint-disable-line no-magic-numbers */

/**
 * @param {Array.<Array.<object>>} groupedSlots grouped slots
 * @returns {*} dropdown
 * @constructor
 */
const ToolbarDropdown = groupedSlots => <BlockFormatControls>
	{ groupedSlots.map( ( slots, groupIndex ) => <div key={ groupIndex } className="editor-format-toolbar block-editor-format-toolbar">
		<Toolbar>
			{ slots.map( ( { Slot, setting }, index ) => <Slot key={ `${ groupIndex }-${ index }` }>
				{ fills => ! fills.length ? null : createComponent( fills, setting ) }
			</Slot> ) }
		</Toolbar>
	</div> ) }
</BlockFormatControls>;

export default ToolbarDropdown;
