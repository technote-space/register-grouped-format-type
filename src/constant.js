export const PLUGIN_NAME = 'register-grouped-format-type';
