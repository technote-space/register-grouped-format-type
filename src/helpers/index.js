export {registerGroupedFormatType, registerFormatTypeGroup} from './toolbar-button';
export {registerMultipleClassFormatType, getFormatName, getRemoveFormatButton} from './rich-text';
