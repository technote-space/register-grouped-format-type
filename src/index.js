import * as RichText from './helpers';

export {RichText};
