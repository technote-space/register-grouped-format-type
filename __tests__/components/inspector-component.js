/* eslint-disable no-magic-numbers */
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import { InspectorComponent, GroupedControls } from '../../src/components';

const { BlockEdit, InspectorControls } = wp.blockEditor;
const { SlotFillProvider, ToolbarButton } = wp.components;
const { Fragment } = wp.element;
const { addFilter, removeFilter } = wp.hooks;
const { create } = wp.richText;

let filter;
beforeAll( () => {
	addFilter( 'editor.BlockEdit', 'components-test/inspector-component', BlockEdit => props => filter( BlockEdit, props ) );
} );

afterAll( () => {
	removeFilter( 'editor.BlockEdit', 'components-test/inspector-component' );
} );

describe( 'InspectorComponent', () => {
	const getSnapshotName = ( name, index ) => `${ name }--${ index }`;
	const setting = index => ( {
		icon: 'admin-customizer',
		position: 'top right',
		label: `test label ${ index }`,
		menuLabel: `test menu label ${ index }`,
		className: undefined,
		menuClassName: undefined,
		inspectorSettings: {},
	} );
	const { Fill, Slot } = GroupedControls( 'components-test' );

	const createFilter = ( index, createToolbarDropdown, createComponents ) => ( BlockEdit, props ) => {
		if ( ! props.isSelected ) {
			return <BlockEdit { ...props }/>;
		}

		return <Fragment>
			<BlockEdit { ...props }/>
			{ createToolbarDropdown( Slot, setting( index ), index ) }
			{ createComponents( Fill, index ) }
		</Fragment>;
	};

	[
		{
			createComponents: Fill => <Fragment>
				<Fill>
					<a href='http://example.com/test1'>test1</a>
				</Fill>
				<Fill>
					<a href='http://example.com/test2'>test2</a>
				</Fill>
			</Fragment>,
			createToolbarDropdown: ( Slot, setting ) => InspectorComponent( Slot, setting ),
			callback: ( wrapper ) => {
				expect( wrapper.find( 'a' ).hostNodes() ).toHaveLength( 2 );
				expect( wrapper.find( '.editor-format-toolbar' ).hostNodes() ).toHaveLength( 0 );
			},
		},
		{
			createComponents: Fill => <Fragment>
				<Fill>
					<ToolbarButton
						title={ 'ToolbarButton1' }
						className={ 'test1' }
						isDisabled={ true }
					/>
				</Fill>
			</Fragment>,
			createToolbarDropdown: ( Slot, setting ) => InspectorComponent( Slot, setting ),
			callback: ( wrapper ) => {
				expect( wrapper.find( 'a' ).hostNodes() ).toHaveLength( 0 );
				expect( wrapper.find( '.editor-format-toolbar' ).hostNodes() ).toHaveLength( 0 );
				expect( wrapper.find( '.test1' ).hostNodes() ).toHaveLength( 0 );
			},
		},
	].forEach( ( { createComponents, createToolbarDropdown, callback }, index ) => {
		it( `should render InspectorComponent ${ index }`, () => {
			filter = createFilter( index, createToolbarDropdown, createComponents );
			const wrapper = mount(
				<SlotFillProvider>
					<InspectorControls.Slot/>

					<BlockEdit
						name="core/paragraph"
						isSelected={ true }
						attributes={ ( {
							className: 'test-block-edit',
							content: create( {
								text: 'test',
								start: 0,
								end: 1,
								formats: [ [], [], [], [] ],
							} ),
						} ) }
					/>
				</SlotFillProvider>,
			);

			expect( toJson( wrapper, { mode: 'deep' } ) ).toMatchSnapshot( getSnapshotName( 'test', index ) );

			callback( wrapper, index );
		} );
	} );
} );
