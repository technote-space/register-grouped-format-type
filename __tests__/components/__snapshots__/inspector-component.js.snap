// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InspectorComponent should render InspectorComponent 0: test--0 1`] = `
Array [
  <div
    className="components-panel__body inspector-test is-opened"
  >
    <a
      href="http://example.com/test1"
      key="1---0/.0"
    >
      test1
    </a>
    <a
      href="http://example.com/test2"
      key="2---0/.0"
    >
      test2
    </a>
    <div
      className="additional-inspector-test1"
      key="additional-inspector-0"
    >
      test1
    </div>
    <div
      className="additional-inspector-test2"
      key="additional-inspector-1"
    >
      test2
    </div>
  </div>,
  Array [
    <div
      onMouseMove={[Function]}
    >
      <div
        className="block-editor-block-list__layout is-root-container"
        onDragStart={[Function]}
        onFocus={[Function]}
      >
        <div
          className="block-list-appender"
          onFocus={[Function]}
          tabIndex={-1}
        >
          <div
            className="wp-block block-editor-default-block-appender"
            data-root-client-id=""
          >
            <textarea
              aria-label="Add block"
              className="block-editor-default-block-appender__content"
              onChange={[Function]}
              onFocus={[Function]}
              readOnly={true}
              role="button"
              rows={1}
              value="Start writing or type / to choose a block"
            />
            <div
              className="components-dropdown block-editor-inserter"
            >
              <button
                aria-expanded={false}
                aria-haspopup="true"
                aria-label="Add block"
                className="components-button block-editor-inserter__toggle has-icon"
                disabled={false}
                onBlur={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onMouseDown={[Function]}
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
                type="button"
              >
                <svg
                  aria-hidden={true}
                  focusable={false}
                  height={24}
                  role="img"
                  viewBox="0 0 24 24"
                  width={24}
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M18 11.2h-5.2V6h-1.6v5.2H6v1.6h5.2V18h1.6v-5.2H18z"
                  />
                </svg>
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>,
    Array [
      "",
      "",
    ],
    Array [
      "",
      "",
    ],
    Array [
      "",
      "",
    ],
  ],
]
`;

exports[`InspectorComponent should render InspectorComponent 1: test--1 1`] = `
Array [
  <div
    className="components-panel__body inspector-test"
  />,
  Array [
    <div
      onMouseMove={[Function]}
    >
      <div
        className="block-editor-block-list__layout is-root-container"
        onDragStart={[Function]}
        onFocus={[Function]}
      >
        <div
          className="block-list-appender"
          onFocus={[Function]}
          tabIndex={-1}
        >
          <div
            className="wp-block block-editor-default-block-appender"
            data-root-client-id=""
          >
            <textarea
              aria-label="Add block"
              className="block-editor-default-block-appender__content"
              onChange={[Function]}
              onFocus={[Function]}
              readOnly={true}
              role="button"
              rows={1}
              value="Start writing or type / to choose a block"
            />
            <div
              className="components-dropdown block-editor-inserter"
            >
              <button
                aria-expanded={false}
                aria-haspopup="true"
                aria-label="Add block"
                className="components-button block-editor-inserter__toggle has-icon"
                disabled={false}
                onBlur={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onMouseDown={[Function]}
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
                type="button"
              >
                <svg
                  aria-hidden={true}
                  focusable={false}
                  height={24}
                  role="img"
                  viewBox="0 0 24 24"
                  width={24}
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M18 11.2h-5.2V6h-1.6v5.2H6v1.6h5.2V18h1.6v-5.2H18z"
                  />
                </svg>
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>,
    Array [
      "",
      "",
    ],
    Array [
      "",
      "",
    ],
    Array [
      "",
      "",
    ],
  ],
]
`;

exports[`InspectorComponent should render InspectorComponent 2: test--2 1`] = `
Array [
  "",
  Array [
    <div
      onMouseMove={[Function]}
    >
      <div
        className="block-editor-block-list__layout is-root-container"
        onDragStart={[Function]}
        onFocus={[Function]}
      >
        <div
          className="block-list-appender"
          onFocus={[Function]}
          tabIndex={-1}
        >
          <div
            className="wp-block block-editor-default-block-appender"
            data-root-client-id=""
          >
            <textarea
              aria-label="Add block"
              className="block-editor-default-block-appender__content"
              onChange={[Function]}
              onFocus={[Function]}
              readOnly={true}
              role="button"
              rows={1}
              value="Start writing or type / to choose a block"
            />
            <div
              className="components-dropdown block-editor-inserter"
            >
              <button
                aria-expanded={false}
                aria-haspopup="true"
                aria-label="Add block"
                className="components-button block-editor-inserter__toggle has-icon"
                disabled={false}
                onBlur={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onMouseDown={[Function]}
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
                type="button"
              >
                <svg
                  aria-hidden={true}
                  focusable={false}
                  height={24}
                  role="img"
                  viewBox="0 0 24 24"
                  width={24}
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M18 11.2h-5.2V6h-1.6v5.2H6v1.6h5.2V18h1.6v-5.2H18z"
                  />
                </svg>
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>,
    "",
    Array [
      "",
      "",
    ],
  ],
]
`;
